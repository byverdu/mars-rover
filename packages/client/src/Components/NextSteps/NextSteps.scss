.next-steps {
  .key-button {
    background-image: url(http://pngimages.net/sites/default/files/keyboard-key-png-image-20213.png);
    width: 50px;
    height: 50px;
    background-size: contain;
    background-repeat: no-repeat;
    border: 0;
    outline: none;
    font-size: 25px;
    transform: scale(1);
    transition: all 1s;
    cursor: pointer;
    padding-bottom: 12px;

    &:active {
      transform: scale(0.8);
    }

    &:disabled {
      cursor: default;
    }
  }
}
