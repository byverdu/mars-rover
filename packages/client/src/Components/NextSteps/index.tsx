import React from 'react';
import './NextSteps.scss';
import { EnumsSteps } from '../../Models/enums';
import { convertSourceInToCoords } from '../../config/utils';

export interface NextStepsProps {
  setRoverNextSteps: (value: string) => void;
  initialPosition: string;
  outOfBoundaries: number;
}

interface NextStepsState {
  steps: string;
  nextPosition: string;
}

export default class NextSteps extends React.PureComponent<
  NextStepsProps,
  NextStepsState
> {
  constructor(props) {
    super(props);

    this.state = {
      steps: '',
      nextPosition: ''
    };
  }

  componentWillReceiveProps(props: NextStepsProps, nextProps: NextStepsProps) {
    if (this.state.steps.length === 0) {
      this.isMoveForwardOutOfBoundaries(
        props.initialPosition,
        props.outOfBoundaries
      );
    }
    if (props.initialPosition === '') {
      this.setState({
        steps: ''
      });
    }
  }

  isMoveForwardOutOfBoundaries(initialPosition, outOfBoundaries) {
    const isMoveForwardOutOfBoundaries = convertSourceInToCoords(
      initialPosition,
      'M',
      outOfBoundaries
    );

    (document.querySelector('[data-step="M"]') as HTMLButtonElement).disabled =
      isMoveForwardOutOfBoundaries === null;
  }

  onClickHandler = (e) => {
    const { initialPosition, setRoverNextSteps, outOfBoundaries } = this.props;
    const { steps, nextPosition } = this.state;

    const elementText = (e.target as HTMLButtonElement).textContent;
    const newText = steps.concat(elementText);
    const thisPosition =
      nextPosition.length !== 0 ? nextPosition : initialPosition;
    const nextCoords = convertSourceInToCoords(
      thisPosition,
      elementText,
      outOfBoundaries
    );

    if (steps.length > 0) {
      this.isMoveForwardOutOfBoundaries(nextCoords, outOfBoundaries);
    }

    this.setState(
      {
        nextPosition: nextCoords,
        steps: newText
      },
      () => setRoverNextSteps(newText)
    );
  };

  render() {
    const { steps } = this.state;

    return (
      <div>
        <h5>Set Rover Next Steps</h5>
        {steps}
        {[EnumsSteps.L, EnumsSteps.M, EnumsSteps.R].map((step) => {
          return (
            <button key={step} data-step={step} onClick={this.onClickHandler}>
              {step}
            </button>
          );
        })}
      </div>
    );
  }
}
