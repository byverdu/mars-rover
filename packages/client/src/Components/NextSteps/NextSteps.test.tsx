import React from 'react';
import NextSteps, { NextStepsProps } from './index';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

const props: NextStepsProps = {
  setRoverNextSteps: (value: string) => {},
  initialPosition: '0 0 N',
  outOfBoundaries: 1
};

describe('<NextSteps />', () => {
  const wrapper = mount(<NextSteps {...props} />);

  it('should create a snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
