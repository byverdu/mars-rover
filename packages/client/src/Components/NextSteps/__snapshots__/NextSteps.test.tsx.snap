// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NextSteps /> should create a snapshot 1`] = `
<NextSteps
  initialPosition="0 0 N"
  outOfBoundaries={1}
  setRoverNextSteps={[Function]}
>
  <div>
    <h5>
      Set Rover Next Steps
    </h5>
    <button
      data-step="L"
      key="L"
      onClick={[Function]}
    >
      L
    </button>
    <button
      data-step="M"
      key="M"
      onClick={[Function]}
    >
      M
    </button>
    <button
      data-step="R"
      key="R"
      onClick={[Function]}
    >
      R
    </button>
  </div>
</NextSteps>
`;
